### PR TITLE
ETSModel with auto-fitting capability similar to R implementation

### LINKED ISSUE #327: Implement Automatic Exponential Smoothing
**Is your feature request related to a problem? Please describe.**

Add the Automatic Exponential Smoothing feature to BaseSktimeForecaster.

**Describe the solution you'd like**

Translate [the fable version on R](https://github.com/tidyverts/fable/blob/master/R/ets.R) to Python and add to sktime.

**Describe alternatives you've considered**

There might be existing Python packages that implement this feature.

**Additional context**

R documentation of the [ETS function](https://www.rdocumentation.org/packages/fable/versions/0.0.0.9100/topics/ETS).

### PR DESCRIPTION
**Pull request by HYang1996**
#### Reference Issues/PRs

Fixes #327, replacing #350 .

#### What does this implement/fix? Explain your changes.

Created a wrapper for statsmodels' latest implementation of ETSModel, and added an auto-fitting option which is similar to the implementation in R.

#### PR checklist for new estimators 

- [x] I've added unit tests and made sure they pass locally. 
- [x] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.
- [x] I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/master/CODEOWNERS) and I'm committed to maintain my algorithm.
- [x] I've added myself to the list of contributors
- [x] I've added the estimator to the online documentation.

#### Any other comments?

Simple tests have been performed to compare results from this forecaster against those from the statsmodels' version, and auto-fitting capability is tested against R implementation on the airline dataset.

### CODE REVIEW

**Review by mloning — COMMENTED** [review]
Hi @HYang1996, thanks for the great work! Code is very clean and readable! 

* the continuous integration tests fail at the moment, not entirely sure what the problem is
* I think we need to update the statsmodels version to `0.12.x` here: [binder](https://github.com/alan-turing-institute/sktime/blob/master/.binder/requirements.txt), [build](https://github.com/alan-turing-institute/sktime/blob/master/build_tools/requirements.txt) and [setup.py](https://github.com/alan-turing-institute/sktime/blob/6557fd8aee0b0fbdb6667413254a820e37050598/setup.py#L22)
* could you add some unit tests where we check results against R's values, I'm happy if you just copy and paste results from R and check if our version returns more or less the same values, the `np.testing` module has some good tools do to that)
* Just another idea: I recently came across this function for storing attributes: https://fastcore.fast.ai/utils.html#store_attr perhaps we could use something similar, that would make it easier to maintain the init functions of all our estimators

File: sktime/forecasting/ets.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+            if self.allow_multiplicative_trend:
+                trend_range = ['add', 'mul', None]
+            else:
+                trend_range = ['add', None]
+            seasonal_range = ['add', 'mul', None]
+            damped_range = [True, False]
+
+            # Fit model, adapted from:
+            # https://github.com/robjhyndman/forecast/blob/master/R/ets.R
+            for error in error_range:
```
Comment (mloning) [inline sktime/forecasting/ets.py:220]: This is probably something we want to parallelise. You could try to make this into a single for loop using [`itertools.product()`](https://docs.python.org/2/library/itertools.html#itertools.product) and joblib.

**Comment by mloning** [comment]
Seems like the CI servers time out because we hit the one hour mark. I may have to speed them up a bit.

**Comment by mloning** [comment]
Okay, now they seem to work, just restarted them. Not sure, I believe Azure had some maintenance work going on over the last few days, perhaps that caused the slow downs.

File: sktime/forecasting/ets.py (hunk 1 above)
Comment (mloning) [reply sktime/forecasting/ets.py:220]:
```python
from itertools import product
from joblib import delayed, Parallel

def _iter(x, y):
    for a, b in product(x, y):
        if a == 1:
            continue
        yield a, b

def _fit(a, b):
    print(a, b)
    return a + b

x = [1, 2, 3]
y = [4, 5, 6]

results = Parallel(n_jobs=1)(delayed(_fit)(a, b) for a, b in _iter(x, y))
```

**Comment by mloning** [comment]
The failure now seems to be related to the changed module structure in statsmodels.

**Comment by mloning** [comment]
Failing docs due to pmdarima and statsmodels )0.12. incompatibility due to this bug: https://github.com/statsmodels/statsmodels/issues/7019

**Comment by HYang1996** [comment]
Unit tests passed locally but failed in checks since rpy2 is not included. The test importing dataset from R is therefore removed for now.

**Comment by HYang1996** [comment]
I'm not exactly sure why the pre-commit check is failing.

File: docs/source/api_reference.rst [hunk 2]
```diff
@@ -200,6 +200,17 @@ Exponential Smoothing
 
     ExponentialSmoothing
 
+ETS Models
```
Comment (mloning) [inline docs/source/api_reference.rst:203]: Shall we not add it to the family of exponential smoothing models?

File: sktime/forecasting/tests/test_ets.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    forecaster = AutoETS(autofitting=True, sp=12, n_jobs=-1,
+                         allow_multiplicative_trend=True)
+    forecaster.fit(y)
+    fitted_forecaster = forecaster._fitted_forecaster
+    fit_result = [fitted_forecaster.error,
+                  fitted_forecaster.trend,
+                  fitted_forecaster.seasonal]
+
+    assert_array_equal(fit_result_R, fit_result)
```
Comment (mloning) [inline sktime/forecasting/tests/test_ets.py:42]: I'd make this a separate function/test, makes it easier to debug if one of them fails

File: sktime/forecasting/ets.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+        dates=None,
+        freq=None,
+        missing="none",
+        start_params=None,
+        maxiter=1000,
+        full_output=True,
+        disp=True,
+        callback=None,
+        return_params=False,
+        autofitting=False,
```
Comment (mloning) [inline sktime/forecasting/ets.py:171]: I'd just call it "auto" rather than "auto fitting"

File: sktime/forecasting/ets.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+    References
+    ----------
+    [1] Hyndman, R.J., & Athanasopoulos, G. (2019) *Forecasting:
+        principles and practice*, 3rd edition, OTexts: Melbourne,
+        Australia. OTexts.com/fpp3. Accessed on April 19th 2020.
+
+    The following parameters are adapted from the ones on R
+    (https://www.rdocumentation.org/packages/forecast/versions/8.12/topics/ets),
+    and are used for automatic model selection:
+
```
Comment (mloning) [inline sktime/forecasting/ets.py:120]: I would move the section `References` to the end of the docstring, I see that you want to reference the different parameters to the different implementations, but I think it's okay to just mention in the references that they come from both model implementations

File: sktime/forecasting/ets.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+                    self.information_criterion != 'bic' and \
+                    self.information_criterion != 'aicc':
+                raise ValueError('information criterion must '
+                                 'either be aic, bic or aicc')
+
+            # Fit model, adapted from:
+            # https://github.com/robjhyndman/forecast/blob/master/R/ets.R
+
+            # Initialise iterator
+            def _iter(error_range, trend_range, seasonal_range, damped_range):
```
Comment (mloning) [inline sktime/forecasting/ets.py:239]: nice, looks good!

File: sktime/forecasting/ets.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+
+            # Fit models
+            _fitted_results = Parallel(n_jobs=self.n_jobs)(
+                delayed(_fit)(error, trend, seasonal, damped)
+                for error, trend, seasonal, damped in _iter(
+                        error_range, trend_range, seasonal_range, damped_range
+                    ))
+
+            # Select best model based on information criterion
+            for result in _fitted_results:
```
Comment (mloning) [inline sktime/forecasting/ets.py:298]: We can do this in a loop, but I wonder if there's a faster way using `np.arrray` where we can simply call something like `np.argmax(_ics)` to find the index of the model with the highest IC, but it's always a bit tricky to get the returned list (`_fitted_results`) into an array ...

File: sktime/forecasting/ets.py [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+                )
+                _fitted_forecaster = _forecaster.fit(
+                    start_params=self.start_params,
+                    maxiter=self.maxiter,
+                    full_output=self.full_output,
+                    disp=self.disp,
+                    callback=self.callback,
+                    return_params=self.return_params
+                )
+                return _forecaster, _fitted_forecaster
```
Comment (mloning) [inline sktime/forecasting/ets.py:288]: Why do we return the unfitted `_forecaster` here?

File: sktime/forecasting/ets.py [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+            self._fitted_forecaster = self._forecaster.fit(
+                start_params=self.start_params,
+                maxiter=self.maxiter,
+                full_output=self.full_output,
+                disp=self.disp,
+                callback=self.callback,
+                return_params=self.return_params
+            )
+
+    def summary(self):
```
Comment (mloning) [inline sktime/forecasting/ets.py:337]: Since this is a method exposed to users, could you add a docstring?

**Review by mloning — COMMENTED** [review]
@HYang1996 looks great! I left a few minor comments, so it's almost good to go now! 

Don't worry about the pre-commit CI checks, I changed the other day and we can ignore them on this PR. If you want to find out more on how they work, check out the [CONTRIBUTING.md](https://github.com/alan-turing-institute/sktime/blob/master/CONTRIBUTING.md#Continuous-integration)

**Comment by HYang1996** [comment]
AutoETS doesn't seem to be appearing in the online documentation.

**Comment by HYang1996** [comment]
> AutoETS doesn't seem to be appearing in the online documentation.

Fixed

**Review by mloning — COMMENTED** [review]
Hi @HYang1996 

Looks great! I'm very happy with it, excellent work! 

One last question: Currently, the unit tests only check if we select the right kind of parameters but not exact numerical values. Is this something we could do using the results from R even if they are just approximately equal?

**Review by mloning — APPROVED** [review]
Excellent work, thanks a lot @HYang1996! 

I'll leave the PR open for a bit longer before I merge in case others want to comment.

**Review by mloning — APPROVED** [review]

**Comment by mloning** [comment]
Thanks again @HYang1996, now merged! 🎉

**Pull request merged**